### PR TITLE
Introduce dataonly for the pkgconfig module

### LINKED ISSUE #4011: pkgconfig module should allow nullification of some default variables.
Not all projects require the prefix, libdir, and includedir variables in their pkgconfig files.
In addition, it should be possbile to remove the `Cflags: -I${includedir}` line either entirely
or in order to have your own variables there, minus the `-I${includedir}` addition.
For example's sake, this is the entireity of the xorg-libinput.pc file:
```
sdkdir=/usr/include/xorg

Name: xorg-libinput
Description: X.Org libinput input driver.
Version: 0.28.0
Cflags: -I${sdkdir}
```

### PR DESCRIPTION
**Pull request by shadeslayer**
This allows users to disable writing out the inbuilt variables to
the pkg-config file as they might actualy not be required.

https://gitlab.freedesktop.org/wayland/weston/issues/269 is a
great example of this.

Fixes #4011

### CODE REVIEW

**Comment by xclaesse** [comment]
You need to also remove the `Cflags: -I${includedir}` in that case since `includedir` won't be defined. I personally don't like the name, what about `common_directories: false`? Dunno... naming is often the hardest part.

**Comment by shadeslayer** [comment]
@xclaesse I have no strong opinions on names tbh, so I'm open to any suggestions that pop up over here.

And yep naming is often the hardest part :(

File: mesonbuild/modules/pkgconfig.py [hunk 1]
```diff
@@ -395,13 +402,19 @@ def generate(self, state, args, kwargs):
         deps.add_priv_reqs(kwargs.get('requires_private', []))
         deps.add_cflags(kwargs.get('extra_cflags', []))
 
+        for dir in subdirs:
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__ [A001] "dir" is a python builtin and is being shadowed, consider renaming the variable ([link](https://sider.review/gh/repos/19784232/pulls/5775#issue-4419191))

<sub>posted by Sider</sub>

**Comment by nirbheek** [comment]
I'd say that the feature actually being exposed is `this pkg-config file is not for a library, but for data`, so I think it'd be better to have a kwarg called `dataonly: (true|false)`, which would only set the data variables. As a bonus, when that's set, setting any kwarg like, `extra_cflags:`, `libraries:`, etc, should result in an error.

File: mesonbuild/modules/pkgconfig.py [hunk 2]
```diff
@@ -368,7 +372,16 @@ def generate(self, state, args, kwargs):
         elif len(args) > 1:
             raise mesonlib.MesonException('Too many positional arguments passed to Pkgconfig_gen.')
 
-        subdirs = mesonlib.stringlistify(kwargs.get('subdirs', ['.']))
+        dataonly = kwargs.get('dataonly', False)
+        if dataonly:
+            default_subdirs = []
+            if len(kwargs.get('libraries'), []) +
+               len(kwargs.get('libraries_private'), []) +
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__ [E113] unexpected indentation ([link](https://sider.review/gh/repos/19784232/pulls/5775#issue-4435334))

<sub>posted by Sider</sub>

File: mesonbuild/modules/pkgconfig.py [hunk 3]
```diff
@@ -368,7 +372,16 @@ def generate(self, state, args, kwargs):
...
+               len(kwargs.get('requires_private'), []) +
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__ [E111] indentation is not a multiple of four ([link](https://sider.review/gh/repos/19784232/pulls/5775#issue-4435335))

<sub>posted by Sider</sub>

File: mesonbuild/modules/pkgconfig.py [hunk 4]
```diff
@@ -368,7 +372,16 @@ def generate(self, state, args, kwargs):
...
+               len(kwargs.get('extra_cflags', [])) > 0:
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__ [E111] indentation is not a multiple of four ([link](https://sider.review/gh/repos/19784232/pulls/5775#issue-4435336))

<sub>posted by Sider</sub>

File: mesonbuild/modules/pkgconfig.py (hunk 2 above)
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__ [E111] indentation is not a multiple of four ([link](https://sider.review/gh/repos/19784232/pulls/5775#issue-4435333))

<sub>posted by Sider</sub>

File: mesonbuild/modules/pkgconfig.py [hunk 5]
```diff
@@ -368,7 +372,16 @@ def generate(self, state, args, kwargs):
         elif len(args) > 1:
             raise mesonlib.MesonException('Too many positional arguments passed to Pkgconfig_gen.')
 
-        subdirs = mesonlib.stringlistify(kwargs.get('subdirs', ['.']))
+        dataonly = kwargs.get('dataonly', False)
+        if dataonly:
+            default_subdirs = []
+            if len(kwargs.get('libraries'), []) +
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__ [E999] SyntaxError: invalid syntax ([link](https://sider.review/gh/repos/19784232/pulls/5775#issue-4435332))

<sub>posted by Sider</sub>

File: mesonbuild/modules/pkgconfig.py [hunk 6]
```diff
@@ -328,23 +330,24 @@ def generate_libs_flags(libs):
...
+                    cflags_buf.append(self._escape(f))
+                return cflags_buf
+
+            cflags = generate_compiler_flags(subdirs)
+            if cflags:
+                ofile.write('Cflags: {}\n'.format(' '.join(cflags)))
 
     @FeatureNewKwargs('pkgconfig.generate', '0.42.0', ['extra_cflags'])
     @FeatureNewKwargs('pkgconfig.generate', '0.41.0', ['variables'])
+    @FeatureNewKwargs('pkgconfig.generate', '0.51.2', ['dataonly'])
```
Comment (nirbheek) [inline mesonbuild/modules/pkgconfig.py]: This should be 0.52.0

File: mesonbuild/modules/pkgconfig.py [hunk 7]
```diff
@@ -418,14 +436,15 @@ def parse_variable_list(stringlist):
                 if any(c.isspace() for c in name):
                     raise mesonlib.MesonException('Invalid whitespace in assignment "{}"'.format(var))
 
-                if name in reserved:
+                if not dataonly and name in reserved:
```
Comment (nirbheek) [inline mesonbuild/modules/pkgconfig.py]: Hmm, why this change?

File: mesonbuild/modules/pkgconfig.py [hunk 8]
```diff
@@ -395,13 +407,19 @@ def generate(self, state, args, kwargs):
         deps.add_priv_reqs(kwargs.get('requires_private', []))
         deps.add_cflags(kwargs.get('extra_cflags', []))
 
+        for d in subdirs:
+            if d == '.':
+                deps.add_cflags(['-I${includedir}'])
+            else:
+                deps.add_cflags(self._escape(PurePath('-I${includedir}') / d))
```
Comment (nirbheek) [inline mesonbuild/modules/pkgconfig.py]: This changes the order of `add_cflags` now, which will cause `-I${includedir}` to appear *after* `extra_cflags`, which is not what we want.

File: mesonbuild/modules/pkgconfig.py [hunk 9]
```diff
@@ -328,23 +330,24 @@ def generate_libs_flags(libs):
...
-            ofile.write('\n')
+
+            def generate_compiler_flags(dirs):
+                cflags_buf = []
+                for f in deps.cflags:
+                    cflags_buf.append(self._escape(f))
+                return cflags_buf
+
+            cflags = generate_compiler_flags(subdirs)
+            if cflags:
```
Comment (nirbheek) [inline mesonbuild/modules/pkgconfig.py]: Shouldn't this be conditional on `not dataonly`?

File: mesonbuild/modules/pkgconfig.py [hunk 10]
```diff
@@ -328,23 +330,24 @@ def generate_libs_flags(libs):
                 ofile.write('Libs: {}\n'.format(' '.join(generate_libs_flags(deps.pub_libs))))
             if len(deps.priv_libs) > 0:
                 ofile.write('Libs.private: {}\n'.format(' '.join(generate_libs_flags(deps.priv_libs))))
```
Comment (nirbheek) [inline mesonbuild/modules/pkgconfig.py]: I think these should also be conditional on `dataonly`, and having any library deps here should be an error. A dataonly pkg-config file can't really pull in libraries, that makes no sense semantically.

File: mesonbuild/modules/pkgconfig.py (hunk 6 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: Ack

File: mesonbuild/modules/pkgconfig.py (hunk 7 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: Yeah, I suppose this is not required anymore.

File: mesonbuild/modules/pkgconfig.py (hunk 8 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: Ack.

File: mesonbuild/modules/pkgconfig.py (hunk 9 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: I suppose that should be additionally conditional on ```not dataonly```

File: mesonbuild/modules/pkgconfig.py (hunk 10 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: Yep

**Comment by jpakkane** [comment]
Featurewise this is ok. Feel free to merge once the details have been hammered out (and docs and a release snippet).

File: mesonbuild/modules/pkgconfig.py (hunk 10 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: @nirbheek any thoughts?

Comment (nirbheek) [reply mesonbuild/modules/pkgconfig.py]: You're right, this is not needed.

**Comment by nirbheek** [comment]
@shadeslayer I'll merge this once you've pushed your changes that fix these comments, and added [a release note snippet for the new feature](https://mesonbuild.com/Contributing.html#special-procedure-for-new-features) thanks!

File: mesonbuild/modules/pkgconfig.py [hunk 11]
```diff
@@ -328,23 +330,24 @@ def generate_libs_flags(libs):
...
-            ofile.write('\n')
+
+            def generate_compiler_flags(dirs):
+                cflags_buf = []
+                for f in deps.cflags:
+                    cflags_buf.append(self._escape(f))
+                return cflags_buf
+
+            cflags = generate_compiler_flags(subdirs)
+            if not dataonly and clfags:
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__ [F821] undefined name 'clfags' ([link](https://sider.review/gh/repos/19784232/pulls/5775#issue-4491475))

<sub>posted by Sider</sub>

File: mesonbuild/modules/pkgconfig.py (hunk 8 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: From taking another pass at this, we add the include dirs and then have a call to 

```deps.add_cflags(kwargs.get('extra_cflags', []))```

So from a logical reading, we'll get ```-I${includedir}``` and then ```extra_cflags``` ?

Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: Hm, the test corroborates your review. I'll take another pass at this.

File: mesonbuild/modules/pkgconfig.py [hunk 12]
```diff
@@ -330,23 +332,24 @@ def generate_libs_flags(libs):
...
+                    cflags_buf.append(self._escape(f))
+                return cflags_buf
+
+            cflags = generate_compiler_flags(subdirs)
+            if not dataonly and cflags:
+                ofile.write('Cflags: {}\n'.format(' '.join(cflags)))
 
     @FeatureNewKwargs('pkgconfig.generate', '0.42.0', ['extra_cflags'])
     @FeatureNewKwargs('pkgconfig.generate', '0.41.0', ['variables'])
+    @FeatureNewKwargs('pkgconfig.generate', '0.52.0', ['dataonly'])
```
Comment (mensinda) [inline mesonbuild/modules/pkgconfig.py]: This should now be `0.53.0`.

File: docs/markdown/snippets/add_release_note_snippets_here [hunk 13]
```diff
@@ -1,3 +1,19 @@
...
+files in projects which also have architecture-dependent outputs.
+
+```
+pkgg.generate(
+  name : 'libhello_nolib',
+  description : 'A minimalistic pkgconfig file.',
+  version : libver,
+  dataonly: true
+)
+```
```
Comment (xclaesse) [inline docs/markdown/snippets/add_release_note_snippets_here]: Wrong file, you should create a new one in this folder

File: mesonbuild/modules/pkgconfig.py [hunk 14]
```diff
@@ -333,23 +335,24 @@ def generate_libs_flags(libs):
...
-                if h == '.':
-                    ofile.write('-I${includedir}')
-                else:
-                    ofile.write(self._escape(PurePath('-I${includedir}') / h))
-            for f in deps.cflags:
-                ofile.write(' ')
-                ofile.write(self._escape(f))
-            ofile.write('\n')
+
+            def generate_compiler_flags(dirs):
```
Comment (xclaesse) [inline mesonbuild/modules/pkgconfig.py]: dirs is not used used in this function. Also, why changing this code at all?

File: mesonbuild/modules/pkgconfig.py [hunk 15]
```diff
@@ -333,23 +335,24 @@ def generate_libs_flags(libs):
...
+                    cflags_buf.append(self._escape(f))
+                return cflags_buf
+
+            cflags = generate_compiler_flags(subdirs)
+            if not dataonly and cflags:
+                ofile.write('Cflags: {}\n'.format(' '.join(cflags)))
 
     @FeatureNewKwargs('pkgconfig.generate', '0.42.0', ['extra_cflags'])
     @FeatureNewKwargs('pkgconfig.generate', '0.41.0', ['variables'])
+    @FeatureNewKwargs('pkgconfig.generate', '0.53.0', ['dataonly'])
```
Comment (xclaesse) [inline mesonbuild/modules/pkgconfig.py]: 0.54

File: mesonbuild/modules/pkgconfig.py [hunk 16]
```diff
@@ -373,7 +377,15 @@ def generate(self, state, args, kwargs):
         elif len(args) > 1:
             raise mesonlib.MesonException('Too many positional arguments passed to Pkgconfig_gen.')
 
-        subdirs = mesonlib.stringlistify(kwargs.get('subdirs', ['.']))
+        dataonly = kwargs.get('dataonly', False)
+        if dataonly:
+            default_subdirs = []
+            blocked_vars = ['libraries', 'libraries_private', 'require_private', 'extra_cflags', 'subdirs']
+            for var in blocked_vars:
+                if len(kwargs.get(var, [])) > 0:
```
Comment (xclaesse) [inline mesonbuild/modules/pkgconfig.py]: Why not simply `var in kwargs` ?

File: test cases/common/47 pkgconfig-gen/meson.build [hunk 17]
```diff
@@ -52,3 +52,10 @@ pkgg.generate(
   description : 'A minimalistic pkgconfig file.',
   version : libver,
 )
+
+pkgg.generate(
+  name : 'libhello_nolib',
+  description : 'A minimalistic pkgconfig file.',
+  version : libver,
+  dataonly: true
```
Comment (xclaesse) [inline test cases/common/47 pkgconfig-gen/meson.build]: You should verify the content of the .pc file is correct too. There are unit tests in run_unittests.py doing that kind of thing.

File: mesonbuild/modules/pkgconfig.py (hunk 14 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: Honestly, I don't even remember why I changed this at this point. I'll just change it back to the original code I suppose.

Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: So I had a more proper look, and seems like the old code would write out a Cflags field even when there are no Cflags to be added to the PC file. So perhaps that's why I modified the code.

**Comment by nirbheek** [comment]
CI failures seem relevant.

File: mesonbuild/modules/pkgconfig.py (hunk 8 above)
Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: according to last comment this should not be resolved conversation.

File: run_unittests.py [hunk 18]
```diff
@@ -4827,6 +4827,10 @@ def test_pkgconfig_gen(self):
         self.assertEqual(foo_dep.get_pkgconfig_variable('foo', {}), 'bar')
         self.assertPathEqual(foo_dep.get_pkgconfig_variable('datadir', {}), '/usr/data')
 
+        libhello_nolib = PkgConfigDependency('libhello_nolib', env, kwargs)
+        self.assertTrue(libhello_nolib.found())
+        self.assertEquallibhello_nolib.get_link_args(), [])
```
Comment (xclaesse) [inline run_unittests.py]: typo here.

Also test that libhello_nolib.get_compile_args() is empty to be complete.

File: mesonbuild/modules/pkgconfig.py (hunk 8 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: I think I've already fixed this as the tests pass now?

Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: The ordering seems correct to me at least in the current version. Wondering if what @nirbheek pointed here was based on previous versions of the PR.

However, you pass it twice in self._escape(), here and in generate_compiler_flags(). That could lead to double quoting of spaces, as far as I understand.

File: mesonbuild/modules/pkgconfig.py (hunk 14 above)
Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: > dirs is not used used in this function

It's a small nitpicking, but that has not been fixed yet.

File: docs/markdown/snippets/pkgconfig_dataonly.md [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+files in projects which also have architecture-dependent outputs.
+
+```
+pkgg.generate(
+  name : 'libhello_nolib',
+  description : 'A minimalistic pkgconfig file.',
+  version : libver,
+  dataonly: true
+)
+```
```
Comment (xclaesse) [inline docs/markdown/snippets/pkgconfig_dataonly.md]: Also need to be documented in Pkgconfig-module.md

File: mesonbuild/modules/pkgconfig.py (hunk 8 above)
Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: I suppose we can drop the escaping in generate_compiler_flags?

Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: It is needed in generate_compiler_flags() to escape other paths that come from dependencies. I think you should dop it from here.

Comment (shadeslayer) [reply mesonbuild/modules/pkgconfig.py]: Done.

**Comment by xclaesse** [comment]
It looks all good to me now, thanks a lot for your patience. But now it has conflicts it seems...

**Comment by shadeslayer** [comment]
@xclaesse heh yeah, I'll resolve it tomorrow :)
